core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,278 @@
+// application master charm configuration.
+func (w *CharmConfigWatcher) appConfigChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	hashCache, ok := msg.(*hashCache)
+	if !ok {
+		logger.Errorf("programming error; application config message was not of expected type, *hashCache")
+	}
+
+	w.masterSettings = hashCache.config
+	w.checkConfig()
+}
+
+// branchChanged is called when we receive a message to say that a branch has
+// been updated in the cache.
+// If it is the branch that this watcher's unit is tracking,
+// check if the latest config delta warrants a notification.
+func (w *CharmConfigWatcher) branchChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	b, okUnit := msg.(Branch)
+	if !okUnit {
+		logger.Errorf("programming error; branch change message was not of expected type, Branch")
+		return
+	}
+
+	if w.branchName != "" {
+		// If we are tracking a different branch to this one, ignore it.
+		if w.branchName != b.Name() {
+			return
+		}
+	} else {
+		// We do not know if we are tracking this branch, so find out.
+		units := b.AssignedUnits()[w.appName]
+		if len(units) == 0 {
+			return
+		}
+		if !set.NewStrings(units...).Contains(w.unitName) {
+			return
+		}
+		w.branchName = b.Name()
+	}
+
+	w.branchDeltas = b.Config()[w.appName]
+	w.checkConfig()
+}
+
+// branchDeleted is called when we receive a message to say that a branch has
+// been removed from the cache.
+// If this watcher's unit was tracking the branch, clean the branch-based
+// details and check if the resulting settings warrant a notification.
+func (w *CharmConfigWatcher) branchDeleted(topic string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	name, okUnit := msg.(string)
+	if !okUnit {
+		logger.Errorf("programming error; branch deleted message was not of expected type, string")
+		return
+	}
+
+	if w.branchName != name {
+		return
+	}
+
+	// The branch we are tracking was deleted.
+	// One of the following scenarios will ensue:
+	// 1) The branch was aborted and this is the only message we will receive.
+	//    Clearing the branch info and checking whether to notify is fine.
+	//    We will send at most one notification.
+	// 2) The branch was committed and this is the first related message.
+	//    There will be a subsequent message for the master settings change.
+	//    If the branch changes mutate master config, there will be 2
+	//    notifications sent.
+	// 3) The branch was committed and this is the second related message,
+	//    coming after the master settings change message.
+	//    The first message should have resulted in no notification,
+	//    because the new master is the same as master + branch deltas.
+	//    When the branch info is removed config remains unchanged,
+	//	  so there is no notification for the second message wither.
+
+	// TODO (manadart 2019-06-18): A fix for case (2) above would be to change
```

Comment:
I don't think this is particularly bad; it's just not great. I will discuss alternative designs.